streaming/base/sampling.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Functionality relating to sampling."""
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+def get_sampling(samples_per_shard: NDArray[np.int64], choose: int, granularity: int, seed: int,
+                 epoch: int, use_epoch: bool) -> NDArray[np.int64]:
+    """Get how many samples to draw from each shard of the given stream.
+
+    Args:
+        samples_per_shard (NDArray[np.int64]): Array of underlying shard sizes.
+        choose (int): How many samples to draw in total over all shards.
+        granularity (int): How many samples to draw at a time from the same shard.
+        seed (int): Seed for shuffling sampling granules.
+        epoch (int): Which epoch we are sampling for.
+        use_epoch (bool): Whether to factor epoch into the base seed, or use the same seed across
+            epochs.
+
+    Returns:
+        NDArray[np.int64]: Array of ephemeral samples chosen per shard.
+    """
+    if choose < 0:
+        raise ValueError('Choose must be a non-negative integer.')
+
+    if granularity <= 0:
+        raise ValueError('Granularity must be a positive integer.')
+
+    if seed < 0:
+        raise ValueError('Seed must be a non-negative integer.')
+
+    if epoch < 0:
+        raise ValueError('Epoch must be a non-negative integer.')
+
+    # Handle whole integer repeat case.
+    num_samples = sum(samples_per_shard)
+    if not choose % num_samples:
+        return samples_per_shard * choose // num_samples
+
+    # Fractional repeat case.
+
+    # Get the ordering by which we will exhaust the shards.
+    pairs = []  # List of (shard ID, samples to draw).
+    for shard_id, shard_samples in enumerate(samples_per_shard):
+        num_granules = (shard_samples + granularity - 1) // granularity
+        shard_ids = np.full(num_granules, shard_id)
+        counts = np.full(num_granules, granularity)
+        if shard_samples % granularity:
+            counts[-1] = shard_samples % granularity
+        pair = shard_ids, counts
+        pairs.append(pair)
+    shard_ids, counts = zip(*pairs)
+    shard_ids = np.concatenate(shard_ids)
+    counts = np.concatenate(counts)
+    num_granules = len(shard_ids)
```

Comment:
there are some duplicate variable names that make the code harder to understand/follow. could you change those as well as add comments? otherwise lgtm!!

Reply:
This for loop can be shorten with np-ops, and (hopefully) it's faster, e.g., 

  sp = np.array(samples_per_shard)
  ng, n_shards = (sp + granularity -1)//granularity, len(samples_per_shard)
  shard_ids = np.repeat(list(range(n_shards)), ng)
  counts = np.repeat([granularity]*n_shards, ng)
  counts[np.cumsum(ng)-1] = sp - (ng-1) * granularity  # iiuc, only the last element has a different value which is this difference.

Reply:
incorporated your changes, please have another look